secure-computer-systems/02-design-principles.md

```diff
@@ -0,0 +1,100 @@
+
+People are often the weakest link in a system and a system is only as good as its weakest link. Requiring people to cycle between a series of related passwords might not actually be a good idea from a usability or security perspective. Design systems that users can accept and that have reasonable expectations about user behavior.
+
+## Economy of Mechanisms
+
+There are tradeoffs between added features and security. 
+With more features there are more ways to attack the system. If you have **ZERO** features then there is no product, but perfect security. There is a balance that can be achieved between useful features and security assurances that we can achieve.
+
+Windows 10 is an operating system with about 50 million lines of code. How could that possibly be secure? Later in the course there is discussion of a *hypervisor* as a TCB instead of the OS itself. A hypervisor is less feature rich and can be implemented in mere hundreds of thousands of lines of code.
+
+#### misc.
+
+This is called economy of mechanism. I am curious the meaning of *mechanism* vs. *feature*, please submit a pull request if you have a good explanation.
+
+Open design is the idea that the attacker should be able to know how the system works and still be unable to execute an attack. The opposite of security by obscurity.
+
+## Least Privilege
+
+Users are authenticated (their identity is verified) and are only allowed to perform certain actions. A user doesn't have to be a person, applications are also considered users in the sense that they have their own privileges/permissions. 
+
+Think about how you have to give applications permissions to do things like look at your pictures on your phone. What would happen if applications had access to all of your photos by default? That sounds like a privacy/security problem.
+
+Different applications have different permissions, and you only give them the permissions they need. This is the principle of least privilege.
+
+### Related Concepts
+
+**Separation of Privileges** is about having more granular ways of controlling access. It makes sense that an application on your phone can separately request access to your camera and to your microphone. This is granularity.
+
+**Fail-safe Defaults** are about denying access by default. The user has no permissions that are not explicitly granted.
+
+
+## Layered Defense or Defense in Depth
+
+Having multiple layers of defense that are diverse will help protect against hackers.
+
+For an example we return to Ken Thompson's Turing Award lecture and discuss a way of protecting ourself from the tricky bug he described.
+
+1. Obtain two independently developed compilers *A* and *B*. Assume that at most one has the bug. Let S<sub>A</sub> and E<sub>A</sub> be the source and executable for *A*. S<sub>B</sub> and E<sub>B</sub> are the source and executable for *B*.
+2. Compile S<sub>A</sub> with E<sub>A</sub> to create *X*. 
```

Comment:
I have support for [Katex](https://katex.org) in these notes ... you can do something like `$S_A$` and it should render correctly.